### PR TITLE
Gnomonic projection using spherical trigonometry

### PR DESCRIPTION
**Pull request by astronasutarou**
The Gnomonic projection is rewritten using Spherical trigonometry.

The previous implementation of the Gnomonic projection in `analysis/projection` used `arctan2(x, -y)` and `1/sqrt(1-z**2)`. These calculations may be problematic when x ~ 0, y ~ 0, or z ~ 1.0. The z value can be close to unity in short-distance cases.

The current implementation uses spherical trigonometry. The calculation is simplified. ~~Note that the accuracy becomes poor when the telescope points to the poles of the coordinate.~~

### CODE REVIEW

File: warpfield/analysis/projection/gnomonic.py [hunk 1]
```diff
@@ -17,31 +17,29 @@ def rotation_matrix(theta):
...
+def gnomonic_cosr(tel_ra, tel_dec, ra, dec):
+    ''' Calculate cos(r), the cosine of the distance r '''
+    return jnp.sin(tel_dec) * jnp.sin(dec) \
+        + jnp.cos(tel_dec) * jnp.cos(dec) * jnp.cos(ra - tel_ra)
 
 
-def gnomonic_y(tel_ra, tel_dec, ra, dec):
-    ''' Calculate internal spherical coordinate y '''
-    return jnp.cos(dec) * jnp.sin(ra - tel_ra)
+def gnomonic_Rsint(tel_ra, tel_dec, ra, dec):
```
[the remark this answers is not in the document]
Comment (astronasutarou) [reply warpfield/analysis/projection/gnomonic.py:26]: The capital R has a special meaning.

File: warpfield/analysis/projection/gnomonic.py [hunk 2]
```diff
@@ -17,31 +17,29 @@ def rotation_matrix(theta):
...
+    ''' Calculate the projected coordinate x '''
+    return jnp.sin(ra - tel_ra) * jnp.cos(dec) \
+        / gnomonic_cosr(tel_ra, tel_dec, ra, dec)
 
 
-def gnomonic_z(tel_ra, tel_dec, ra, dec):
-    ''' Calculate internal spherical coordinate z '''
-    return jnp.cos(tel_dec) * jnp.cos(dec) * jnp.cos(ra - tel_ra) \
-        + jnp.sin(tel_dec) * jnp.sin(dec)
+def gnomonic_Rcost(tel_ra, tel_dec, ra, dec):
```
[the remark this answers is not in the document]
Comment (astronasutarou) [reply warpfield/analysis/projection/gnomonic.py:32]: The capital R has a special meaning.

File: warpfield/analysis/projection/gnomonic.py [hunk 3]
```diff
@@ -17,31 +17,29 @@ def rotation_matrix(theta):
...
+    ''' Calculate the projected coordinate y '''
+    return (jnp.sin(dec) * jnp.cos(tel_dec) \
```
[the remark this answers is not in the document]
Comment (astronasutarou) [reply warpfield/analysis/projection/gnomonic.py]: Solved.
